logsmanagement/parser.c

```diff
@@ -0,0 +1,1622 @@
+        }
+    }
+
+    if (fQuote) {
+        return -1;
+    }
+
+    return cnt;
+}
+
+/**
+ * @brief Parse a delimited string into an array of strings.
+ * @details Given a string containing no linebreaks, or containing line breaks
+ * which are escaped by "double quotes", extract a NULL-terminated
+ * array of strings, one for every delimiter-separated value in the row.
+ * @param[in] line The input string to be parsed.
+ * @param[in] delimiter The delimiter to be used to split the string.
+ * @param[in] num_fields The expected number of fields in \p line. If a negative
+ * number is provided, they will be counted.
+ * @return A NULL-terminated array of strings with the delimited values in \p line,
+ * or NULL in any other case.
+ * @todo This function is not benchmarked or optimised.
+ */
+static inline char **parse_csv( const char *line, const char delimiter, int num_fields) {
+    char **buf, **bptr, *tmp, *tptr;
+    const char *ptr;
+    int fQuote, fEnd;
+
+    if(num_fields < 0){
+        num_fields = count_fields(line, delimiter);
+
+        if ( num_fields == -1 ) {
+            return NULL;
+        }
+    }
+
+    buf = mallocz( sizeof(char*) * (num_fields+1) );
+
+    if ( !buf ) {
+        return NULL;
+    }
+
+    tmp = mallocz( strlen(line) + 1 );
+
+    if ( !tmp ) {
+        freez( buf );
+        return NULL;
+    }
+
+    bptr = buf;
+
+    for ( ptr = line, fQuote = 0, *tmp = '\0', tptr = tmp, fEnd = 0; ; ptr++ ) {
+        if ( fQuote ) {
+            if ( !*ptr ) {
+                break;
+            }
+
+            if ( *ptr == '\"' ) {
+                if ( ptr[1] == '\"' ) {
+                    *tptr++ = '\"';
+                    ptr++;
+                    continue;
+                }
+                fQuote = 0;
+            }
+            else {
+                *tptr++ = *ptr;
+            }
+
+            continue;
+        }
+
+
+        if(*ptr == '\"'){
+            fQuote = 1;
+            continue;
+        }
+        else if(*ptr == '\0'){
+            fEnd = 1;
+            *tptr = '\0';
+            *bptr = strdupz( tmp );
+
+            if ( !*bptr ) {
+                for ( bptr--; bptr >= buf; bptr-- ) {
+                    freez( *bptr );
+                }
+                freez( buf );
+                freez( tmp );
+
+                return NULL;
+            }
+
+            bptr++;
+            tptr = tmp;
+            break;
+        }
+        else if(*ptr == delimiter){
+            *tptr = '\0';
+            *bptr = strdupz( tmp );
+
+            if ( !*bptr ) {
+                for ( bptr--; bptr >= buf; bptr-- ) {
+                    freez( *bptr );
+                }
+                freez( buf );
+                freez( tmp );
+
+                return NULL;
+            }
+
+            bptr++;
+            tptr = tmp;
+
+            continue;
+        }
+        else{
+            *tptr++ = *ptr;
+            continue;
+        }
+
+        /*switch( *ptr ) {
```

Comment:
Why are you keeping next block commented? Cannot we remove it?

Reply:
I think I was planning to use it for something which I am forgetting now. That function needs rewriting in general.

Reply:
All right, thanks to clarify.